aspnet/fundamentals/routing.rst

```diff
@@ -1,10 +1,288 @@
+
+.. warning:: Avoid using constraints for **validation**, because doing so means that invalid input will result in a 404 (Not Found) instead of a 400 with an appropriate error message. Route constraints should be used to **disambiguate** between routes, not validate the inputs for a particular route.
+
+Constraints can be *chained*. You can specify that a route value is of a certain type and also must fall within a specified range, for example: ``{age:int:range(1,120)}``.  Numeric constraints like ``min``, ``max``, and ``range`` will automatically convert the value to ``long`` before being applied unless another numeric type is specified.
+
+Route templates must be unambiguous, or they will be ignored. For example, ``{id?}/{foo}`` is ambiguous, because it's not clear which route value would be bound to a request for "/bar". Similarly, ``{*everything}/{plusone}`` would be ambiguous, because the first route parameter would match everything from that part of the request on, so it's not clear what the ``plusone`` parameter would match.
+
+.. note:: There is a special case route for filenames, such that you can define a route value like ``files/{filename}.{ext?}``. When both ``filename`` and ``ext`` exist, both values will be populated. However, if only ``filename`` exists in the URL, the trailing period ``.`` is also optional. Thus, these would both match: ``/files/foo.txt`` and ``/files/foo``.
+
+.. tip:: Enable :doc:`logging` to see how the built in routing implementations, such as ``TemplateRoute``, match requests.
+
+Route Builder Extensions
+------------------------
+Several `extension methods on RouteBuilder <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Builder/RouteBuilderExtensions/index.html>`_ are available for convenience. The most common of these is ``MapRoute``, which allows the specification of a route given a name and template, and optionally default values, constraints, and/or data tokens. Data tokens are user defined data that is carried along if the route matches. When using these extensions, you must have specified the ``DefaultHandler`` and ``ServiceProvider`` properties of the ``RouteBuilder`` instance to which you're adding the route. These ``MapRoute`` extensions add new ``TemplateRoute`` instances to the ``RouteBuilder that each target the ``IRouter`` configured as the ``DefaultHandler``.
+
+.. note:: ``MapRoute`` doesn't take an ``IRouter`` parameter - it only adds routes that will be handled by the ``DefaultHandler``. This means that if the default handler is itself a template route with named route value parameters, only these parameters can be used by routes defined using ``MapRoute``. Since the default handler is an ``IRouter``, it may decide not to handle the request. For example, MVC is typically configured as a default handler that only handles requests that match an available controller action.
```

Comment:
Ok, I can pull that out if you like. It was something I included before I started using the DelegateRequestRouter, and at that point the only thing "out of the box" I'd gotten to work with the DefaultHandler was an existing TemplateRoute. In that scenario, if I had a template route like "{foo}/{bar}" as the default, I couldn't then add a new template route of the form "{foo}/{bar}/{baz}" because baz wasn't in the default one, and so would never have a value (when the default route would run).

This probably isn't something most folks will be doing, nor will it be an issue at all once the Delegate router is in the framework, as I believe it is in RC2. Just remove it, do you think?